DOC-10466 - Add Information on Minimum Memory Resident Ratio (#2925)
couchbase/docs-server: modules/manage/pages/manage-buckets/create-bucket.adoc

== Create a Bucket with The UI

To create a bucket with the Couchbase UI: 

. Log in to the Couchbase Server Web Console.
. Go to [.ui]*Buckets*.
. Select btn:[Add Bucket].
+
The [.ui]*Add Data Bucket* dialog appears:

[#add-data-bucket-dialog-initial]
image::manage-buckets/addDataBucketDialogInitial.png[,320,align=left]

[start="4"]
. In the *Name* field, enter a name for the new bucket.

[NOTE]
====
A bucket name can be up to 100 characters in length and contain:

* Uppercase and lowercase characters (A-Z and a-z)
* Digits (0-9)
* Underscores (_), periods (.), dashes (-), and percent symbols (%)
====
[start="5"]
. Choose a *Bucket Type* for the bucket: 
+
* *Couchbase*
* *Memcached*
* *Ephemeral*
+
For more information about bucket types, see xref:learn:buckets-memory-and-storage/buckets.adoc[].
. Choose a *Storage Backend* for the bucket:
+
* *Couchstore*
* *Magma*
+
For more information about the available storage engines, see xref:learn:buckets-memory-and-storage/storage-engines.adoc[Storage Engines].

. In the *Memory Quota* field, enter a value in MiB per node for the total RAM available for the bucket. This value can't exceed the total RAM quota for your cluster. 

> NOTE: Your memory quota needs to match the minimum memory resident ratio required by your chosen storage engine. For more information, see xref:learn:buckets-memory-and-storage/memory.adoc#bucket-memory[Bucket Memory Quotas].

Reviewer: ```suggestion
NOTE: Your memory quota needs to match the minimum memory resident ratio required by your chosen storage engine. 
For more information, see xref:learn:buckets-memory-and-storage/memory.adoc#bucket-memory[Bucket Memory Quotas].
```

For information about how to import documents into a bucket, see xref:manage:import-documents/import-documents.adoc[Import Documents].

[#advanced-bucket-settings]
=== Set Advanced Bucket Settings

The available advanced settings for your bucket change based on your selected *Bucket Type*:

* <<couchbase-bucket-settings>>
* <<memcached-bucket-settings>>
* <<ephemeral-bucket-settings>>

[#couchbase-bucket-settings]
==== Couchbase Bucket Settings

To configure advanced settings for a Couchbase bucket: 

. To enable xref:learn:clusters-and-availability/intra-cluster-replication.adoc[replica creation and management], under *Replicas*, select the btn:[Enable] checkbox. 

.. In the *Number of replica (backup) copies* list, select the number of replicas for the bucket. 

.. To replicate view indexes and data from the bucket, select the btn:[Replicate view indexes] checkbox.

. To set a xref:learn:data/expiration.adoc[document expiration] for documents in the bucket, under *Bucket Max Time-To-Live*, select the btn:[Enable] checkbox.

.. In the *Seconds* field, enter the maximum time in seconds that a document can exist in the bucket before it's deleted. 
+

> TIP: The maximum allowed value is 2147483648 seconds (68.096 years). You can only apply this setting to documents created after you change the configuration. 

Reviewer: ```suggestion
TIP: The maximum allowed value is 2147483648 seconds (68.096 years). 
You can only apply this setting to documents created after you change the configuration. 
```

. Choose a *Compression Mode* for the bucket: 
+
* *Off* 
* *Passive*
* *Active*
+
For more information about the available compression modes, see xref:learn:buckets-memory-and-storage/compression.adoc[].

. Choose a *Conflict Resolution* method for Cross Datacenter Replication (XDCR) on this bucket: 
+
* *Sequence number* 
* *Timestamp*
+
For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].

. Choose an *Ejection Method* for the bucket: 
+
* *Value-only*
* *Full*
+
For more information about ejection, see the xref:learn:buckets-memory-and-storage/memory.adoc#ejection[Ejection section in Memory].
+
include::learn:partial$full-ejection-note.adoc[]

[#bucket-priority, start="6"]
. Choose a *Bucket Priority* for the bucket: 
+
* *Default*
* *High*
+
Bucket Priority sets the priority of the bucket's background tasks relative to the background tasks of other buckets on the cluster. 
+

> Background tasks may involve disk I/O, DCP stream-processing, item-paging, and more. Specifying High might result in faster processing for the current bucket's tasks. This setting only takes effect when there is more than one bucket defined for the cluster, and you have assigned different Bucket Priority values.

Reviewer: ```suggestion
Background tasks may involve disk I/O, DCP stream-processing, item-paging, and more. 
Specifying High might result in faster processing for the current bucket's tasks. 
This setting only takes effect when there is more than one bucket defined for the cluster, and you have assigned different Bucket Priority values.
```

==== Memcached Bucket Settings

CAUTION: Memcached buckets are deprecated. Use a *Couchbase* or *Ephemeral* bucket, instead. 

To configure advanced settings for a Memcached bucket:

. To enable flushing for the bucket, under *Flush*, select the btn:[Enable] checkbox.
+
For more information about flushing, see xref:manage-buckets/flush-bucket.adoc[Flush a Bucket].

[#add-data-bucket-dialog-expanded-for-memcached]
image::manage-buckets/addDataBucketDialogExpandedForMemcached.png[,350,align=left]

[#ephemeral-bucket-settings]
==== Ephemeral Bucket Settings

To configure advanced settings for an Ephemeral bucket: 

. To enable xref:learn:clusters-and-availability/intra-cluster-replication.adoc[replica creation and management], under *Replicas*, select the btn:[Enable] checkbox. 

.. In the *Number of replica (backup) copies* list, select the number of replicas for the bucket. 

. . To set a xref:learn:data/expiration.adoc[document expiration] for documents in the bucket, under *Bucket Max Time-To-Live*, select the btn:[Enable] checkbox.

.. In the *Seconds* field, enter the maximum number of seconds a document can exist in the bucket before it's deleted. 
+

> TIP: The maximum allowed value is 2147483648 seconds (68.096 years). You can only apply this setting to documents created after you change the configuration. 

Reviewer: ```suggestion
TIP: The maximum allowed value is 2147483648 seconds (68.096 years). 
You can only apply this setting to documents created after you change the configuration. 
```

. Choose a *Compression Mode* for the bucket: 
+
* *Off*
* *Passive*
* *Active* 
+
For more information about the available compression modes, see xref:learn:buckets-memory-and-storage/compression.adoc[].

. Choose a *Conflict Resolution* method for Cross Datacenter Replication (XDCR) on this bucket: 
+
* *Sequence number*
* *Timestamp* 
+
For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].

. Choose a *Bucket Priority*: 
+
* *Default*
* *High*
+
Bucket Priority sets the priority of the bucket's background tasks relative to the background tasks of other buckets on the cluster. 
+

> Background tasks may involve DCP stream-processing, item-paging, and more. Specifying High might result in faster processing for the current bucket's tasks. This setting only takes effect when there is more than one bucket defined for the cluster, and the buckets are assigned different Bucket Priority values.

Reviewer: ```suggestion
Background tasks may involve DCP stream-processing, item-paging, and more. 
Specifying High might result in faster processing for the current bucket's tasks. T
his setting only takes effect when there is more than one bucket defined for the cluster, and the buckets are assigned different Bucket Priority values.
```

. Choose an *Ejection Policy* for the bucket: 
+
* *No ejection* 
* *Eject data when RAM is full* 
+
For more information about ejection, see the xref:learn:buckets-memory-and-storage/memory.adoc#ejection[Ejection section in Memory].

. In the *Metadata Purge Interval* field, enter a value between `0.0007-60` to set how often a node purges metadata on deleted items.
+

> A value of `0.0007` equals a minute. A value of `0.5` equals 12 hours. If this value is too high, the node might have a delay when reclaiming memory. If set too low, data might be inconsistent in XDCR or Views. 

Reviewer: ```suggestion
A value of `0.0007` equals a minute. 
A value of `0.5` equals 12 hours. 
If this value is too high, the node might have a delay when reclaiming memory. 
If set too low, data might be inconsistent in XDCR or Views. 
```

= Create a Bucket
:description: pass:q[_Full_ and _Cluster_ Administrators can use Couchbase Web Console, the CLI, or the REST API to create a bucket.]
:page-aliases: clustersetup:create-bucket
:page-topic-type: guide

[abstract]
{description}

You can create a bucket with the xref:manage:manage-buckets/create-bucket.adoc#create-bucket-using-couchbase-web-console[Couchbase Server UI], xref:manage:manage-buckets/create-bucket.adoc#create-bucket-with-the-cli[CLI] or the xref:manage:manage-buckets/create-bucket.adoc#create-bucket-with-the-rest-api[REST API].

You can create a maximum of 30 buckets per cluster.

== Prerequisites

* You must be a Full or Cluster Administrator.

[#create-bucket-using-couchbase-web-console]

> == Create a Bucket with The UI

Reviewer 2: "The" shouldn't be capitalized here

```suggestion
== Create a Bucket with the UI
```

To create a bucket with the Couchbase UI: 

. Log in to the Couchbase Server Web Console.
. Go to [.ui]*Buckets*.
. Select btn:[Add Bucket].
+
The [.ui]*Add Data Bucket* dialog appears:

[#add-data-bucket-dialog-initial]

> image::manage-buckets/addDataBucketDialogInitial.png[,320,align=left]

Reviewer 2: I know you inherited this, but I think screenshots should be center-aligned throughout. Left-aligned screenshots end up  outside the left margin of the text in a multi-level bulleted / numbered list, which visually breaks up the procedure.

(There's a case that other types of images should be left-aligned in other contexts.)

Also consider adding alt text for all screenshots if you have time &mdash; for example, "The Add Data Bucket dialog with initial options"

[start="4"]
. In the *Name* field, enter a name for the new bucket.

[NOTE]
====
A bucket name can be up to 100 characters in length and contain:

* Uppercase and lowercase characters (A-Z and a-z)
* Digits (0-9)
* Underscores (_), periods (.), dashes (-), and percent symbols (%)
====
[start="5"]
. Choose a *Bucket Type* for the bucket: 
+
* *Couchbase*
* *Memcached*
* *Ephemeral*
+
For more information about bucket types, see xref:learn:buckets-memory-and-storage/buckets.adoc[].

> . Choose a *Storage Backend* for the bucket:

Reviewer 2: The paragraph beginning "For more information..." belongs semantically to the numbered step, not to the last bullet point. I think it should be outdented so that it aligns with the text beginning "Choose a **Bucket Type** ..."
To do this, you need to add a single blank line before the continuation mark `+`:

```suggestion
* *Ephemeral*

+
For more information about bucket types, see xref:learn:buckets-memory-and-storage/buckets.adoc[].
. Choose a *Storage Backend* for the bucket:
```

+
* *Couchstore*
* *Magma*
+

> For more information about the available storage engines, see xref:learn:buckets-memory-and-storage/storage-engines.adoc[Storage Engines].

Reviewer 2: Likewise, outdent the "For more information..." paragraph slightly to line it up with "Choose a **Storage Backend** ..."

```suggestion
* *Magma*

+
For more information about the available storage engines, see xref:learn:buckets-memory-and-storage/storage-engines.adoc[Storage Engines].
```

. In the *Memory Quota* field, enter a value in MiB per node for the total RAM available for the bucket. This value can't exceed the total RAM quota for your cluster. 

NOTE: Your memory quota needs to match the minimum memory resident ratio required by your chosen storage engine. 
For more information, see xref:learn:buckets-memory-and-storage/memory.adoc#bucket-memory[Bucket Memory Quotas].

[start="8"]
. Expand *Advanced bucket settings*. 
. Set any advanced settings for your bucket. See <<advanced-bucket-settings>>.
. Select btn:[Add Bucket].

> The bucket appears on the *Buckets* screen: 

Reviewer 2: I recommend not using a colon before a screenshot. If the screenshot is deleted in a future revision, or the paragraph gets moved around, the colon is just one extra thing to remove and it frequently gets left.

image::manage-buckets/bucketsViewWithCreatedBucket.png[,750,align=left]

You can view the following information for the bucket: 

* The number of items the bucket contains. 
* The percentage of items in the bucket that are currently resident. 
* The number of operations per second being performed on the bucket. 

> * The amount of RAM the bucekt is currently using, compared to its memory quota. 

Reviewer 2: Typo

```suggestion
* The amount of RAM the bucket is currently using, compared to its memory quota. 
```

* The amount of disk space used by the bucket. 

You can also view the xref:manage:manage-ui/manage-ui.adoc#console-documents[Documents] in the bucket and create xref:learn:data/scopes-and-collections.adoc[Scopes and Collections]. 

For information about how to import documents into a bucket, see xref:manage:import-documents/import-documents.adoc[Import Documents].

[#advanced-bucket-settings]
=== Set Advanced Bucket Settings

The available advanced settings for your bucket change based on your selected *Bucket Type*:

* <<couchbase-bucket-settings>>
* <<memcached-bucket-settings>>
* <<ephemeral-bucket-settings>>

[#couchbase-bucket-settings]
==== Couchbase Bucket Settings

To configure advanced settings for a Couchbase bucket: 

> . To enable xref:learn:clusters-and-availability/intra-cluster-replication.adoc[replica creation and management], under *Replicas*, select the btn:[Enable] checkbox. 

Reviewer 2: I don't think a check box is a button. It's not something the user clicks to make an action happen straight away. I would just use bold for check boxes throughout.

Author: Yeah I waffled on this a bit just because of the [.ui] labels that were left on all of the checkboxes.

.. In the *Number of replica (backup) copies* list, select the number of replicas for the bucket. 

.. To replicate view indexes and data from the bucket, select the btn:[Replicate view indexes] checkbox.

. To set a xref:learn:data/expiration.adoc[document expiration] for documents in the bucket, under *Bucket Max Time-To-Live*, select the btn:[Enable] checkbox.

.. In the *Seconds* field, enter the maximum time in seconds that a document can exist in the bucket before it's deleted. 
+

> TIP: The maximum allowed value is 2147483648 seconds (68.096 years). 

Reviewer 2: This admonition is part of the substep above (so appears indented) &mdash; which is fair enough, but is inconsistent with the "memory quota" note on line 56, which is outside the flow of the numbered steps. Needs to be consistent either way.

You can only apply this setting to documents created after you change the configuration. 

. Choose a *Compression Mode* for the bucket: 
+
* *Off* 
* *Passive*

> * *Active*

Reviewer 2: Add a single blank line before the continuation symbol `+`, so that "For more information ..." lines up with "Choose a Compression mode ..."

```suggestion
* *Active*

```

+
For more information about the available compression modes, see xref:learn:buckets-memory-and-storage/compression.adoc[].

. Choose a *Conflict Resolution* method for Cross Datacenter Replication (XDCR) on this bucket: 
+
* *Sequence number* 
* *Timestamp*
+

> For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].

Reviewer 2: Outdent "For more information ...." to line up with "Choose a ..."

```suggestion
* *Timestamp*

+
For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].
```

. Choose an *Ejection Method* for the bucket: 
+
* *Value-only*
* *Full*
+

> For more information about ejection, see the xref:learn:buckets-memory-and-storage/memory.adoc#ejection[Ejection section in Memory].

Reviewer 2: Outdent as above

```suggestion
* *Full*

+
For more information about ejection, see the xref:learn:buckets-memory-and-storage/memory.adoc#ejection[Ejection section in Memory].
```

+
include::learn:partial$full-ejection-note.adoc[]

[#bucket-priority, start="6"]
. Choose a *Bucket Priority* for the bucket: 
+
* *Default*
* *High*

> +

Reviewer 2: Outdent as above

```suggestion
* *High*

+
```

Bucket Priority sets the priority of the bucket's background tasks relative to the background tasks of other buckets on the cluster. 
+
Background tasks may involve disk I/O, DCP stream-processing, item-paging, and more. 
Specifying High might result in faster processing for the current bucket's tasks. 
This setting only takes effect when there is more than one bucket defined for the cluster, and you have assigned different Bucket Priority values.

[#durability-level, start="7"]
. In the *Minimum Durability Level* list, select a durability level for the bucket: 
+
* *none*
* *majority*
* *majorityAndPersistActive*
* *persistToMajority*
+

> For more information about durability, see xref:learn:data/durability.adoc[].

Reviewer 2: Outdent as above

```suggestion
* *persistToMajority*

+
For more information about durability, see xref:learn:data/durability.adoc[].
```

+
For more information about flushing, see xref:manage-buckets/flush-bucket.adoc[Flush a Bucket].

[#add-data-bucket-dialog-expanded]
image::manage-buckets/addBucketWithMagmaOption.png[,400,align=left]

[#memcached-bucket-settings]
==== Memcached Bucket Settings

CAUTION: Memcached buckets are deprecated. Use a *Couchbase* or *Ephemeral* bucket, instead. 

To configure advanced settings for a Memcached bucket:

. To enable flushing for the bucket, under *Flush*, select the btn:[Enable] checkbox.
+
For more information about flushing, see xref:manage-buckets/flush-bucket.adoc[Flush a Bucket].

[#add-data-bucket-dialog-expanded-for-memcached]
image::manage-buckets/addDataBucketDialogExpandedForMemcached.png[,350,align=left]

[#ephemeral-bucket-settings]
==== Ephemeral Bucket Settings

To configure advanced settings for an Ephemeral bucket: 

. To enable xref:learn:clusters-and-availability/intra-cluster-replication.adoc[replica creation and management], under *Replicas*, select the btn:[Enable] checkbox. 

.. In the *Number of replica (backup) copies* list, select the number of replicas for the bucket. 

> . . To set a xref:learn:data/expiration.adoc[document expiration] for documents in the bucket, under *Bucket Max Time-To-Live*, select the btn:[Enable] checkbox.

Reviewer 2: Extra full stop?

```suggestion
. To set a xref:learn:data/expiration.adoc[document expiration] for documents in the bucket, under *Bucket Max Time-To-Live*, select the btn:[Enable] checkbox.
```

.. In the *Seconds* field, enter the maximum number of seconds a document can exist in the bucket before it's deleted. 
+
TIP: The maximum allowed value is 2147483648 seconds (68.096 years). 
You can only apply this setting to documents created after you change the configuration. 

. Choose a *Compression Mode* for the bucket: 
+
* *Off*
* *Passive*
* *Active* 
+

> For more information about the available compression modes, see xref:learn:buckets-memory-and-storage/compression.adoc[].

Reviewer 2: Outdent as above

```suggestion
* *Active* 

+
For more information about the available compression modes, see xref:learn:buckets-memory-and-storage/compression.adoc[].
```

. Choose a *Conflict Resolution* method for Cross Datacenter Replication (XDCR) on this bucket: 
+
* *Sequence number*
* *Timestamp* 
+

> For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].

Reviewer 2: Outdent as above

```suggestion
* *Timestamp* 

+
For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].
```

. Choose a *Bucket Priority*: 
+
* *Default*
* *High*
+

> Bucket Priority sets the priority of the bucket's background tasks relative to the background tasks of other buckets on the cluster. 

Reviewer 2: Outdent as above

```suggestion
* *High*

+
Bucket Priority sets the priority of the bucket's background tasks relative to the background tasks of other buckets on the cluster. 
```

+
Background tasks may involve DCP stream-processing, item-paging, and more. 
Specifying High might result in faster processing for the current bucket's tasks. T

> his setting only takes effect when there is more than one bucket defined for the cluster, and the buckets are assigned different Bucket Priority values.

Reviewer 2: Typo

```suggestion
Specifying High might result in faster processing for the current bucket's tasks.
This setting only takes effect when there is more than one bucket defined for the cluster, and the buckets are assigned different Bucket Priority values.
```

. Choose an *Ejection Policy* for the bucket: 
+
* *No ejection* 
* *Eject data when RAM is full* 
+

> For more information about ejection, see the xref:learn:buckets-memory-and-storage/memory.adoc#ejection[Ejection section in Memory].

Reviewer 2: Outdent as above

```suggestion
* *Eject data when RAM is full* 

+
For more information about ejection, see the xref:learn:buckets-memory-and-storage/memory.adoc#ejection[Ejection section in Memory].
```

. In the *Metadata Purge Interval* field, enter a value between `0.0007-60` to set how often a node purges metadata on deleted items.
+
A value of `0.0007` equals a minute. 
A value of `0.5` equals 12 hours. 
If this value is too high, the node might have a delay when reclaiming memory. 
If set too low, data might be inconsistent in XDCR or Views. 

. In the *Minimum Durability Level* list, select a durability level for the bucket: 
+
* *none*
* *majority*
+

> For more information about durability, see xref:learn:data/durability.adoc[].

Reviewer 2: Outdent as above

```suggestion
* *majority*

+
For more information about durability, see xref:learn:data/durability.adoc[].
```

[#add-data-bucket-dialog-expanded-for-ephemeral]
image::manage-buckets/addDataBucketDialogExpandedForEphemeral.png[,350,align=left]

[#create-bucket-with-the-cli]
== Create a Bucket with the CLI

To create a bucket with the Couchbase CLI, use the `bucket-create` command.

For example:

> ----

Reviewer 2: Markup as code so users can copy and paste

```suggestion
[source,sh]
----
```

. To set a xref:learn:data/expiration.adoc[document expiration] for documents in the bucket, under *Bucket Max Time-To-Live*, select the *Enable* checkbox.

.. In the *Seconds* field, enter the maximum number of seconds a document can exist in the bucket before it's deleted. 
+
TIP: The maximum allowed value is 2147483648 seconds (68.096 years). 
You can only apply this setting to documents created after you change the configuration. 

. Choose a *Compression Mode* for the bucket: 
+
* *Off*
* *Passive*
* *Active* 

+
For more information about the available compression modes, see xref:learn:buckets-memory-and-storage/compression.adoc[].

. Choose a *Conflict Resolution* method for Cross Datacenter Replication (XDCR) on this bucket: 
+
* *Sequence number*
* *Timestamp* 

+
For more information about XDCR conflict resolution, see xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc[].

. Choose a *Bucket Priority*: 
+
* *Default*
* *High*

+
Bucket Priority sets the priority of the bucket's background tasks relative to the background tasks of other buckets on the cluster. 

Background tasks may involve DCP stream-processing, item-paging, and more. 

> Specifying High might result in faster processing for the current bucket's tasks. T

Reviewer 2: The typo "T" is still there at the end of the line

```suggestion
Specifying High might result in faster processing for the current bucket's tasks.
```

--enable-flush 0
----

> The preceding example creates a `Couchbase` bucket named `testBucket`, with a RAM size of `1024`. It sets a Maximum Time-to-Live and disables Flush. It also sets a Minimum Durability Level of `persistToMajority`. 

Reviewer: ```suggestion
The preceding example creates a `Couchbase` bucket named `testBucket`, with a RAM size of `1024`.
 It sets a Maximum Time-to-Live and disables Flush. 
 It also sets a Minimum Durability Level of `persistToMajority`. 
```

--enable-flush 0
----

The preceding example creates a `Couchbase` bucket named `testBucket`, with a RAM size of `1024`.
 It sets a Maximum Time-to-Live and disables Flush. 
 It also sets a Minimum Durability Level of `persistToMajority`. 

For more information about `bucket-create` and its parameters, see xref:cli:cbcli/couchbase-cli-bucket-create.adoc[bucket-create] in the Couchbase CLI reference.

[#create-bucket-with-the-rest-api]
== Create a Bucket with the REST API

To create a bucket with the Couchbase REST API, use the `POST` http method, with the `/pools/default/buckets` endpoint.

For example:

> ----

Reviewer 2: Markup as code so users can copy and paste

```suggestion
[source,sh]
----
```

-d durabilityMinLevel=majorityAndPersistActive
----

> The preceding example creates a `Couchbase` bucket named `testBucket`, with a RAM size of `512`. It sets a Minimum Durability Level of `majorityAndPersistActive`.

Reviewer: ```suggestion
The preceding example creates a `Couchbase` bucket named `testBucket`, with a RAM size of `512`. 
It sets a Minimum Durability Level of `majorityAndPersistActive`.
```
